vassal-app/src/main/java/VASSAL/tools/SequenceEncoder.java

```diff
@@ -236,7 +235,13 @@ public boolean hasMoreTokens() {
     }
 
     public String nextToken() {
-      if (!hasMoreTokens()) throw new NoSuchElementException();
+      if (!hasMoreTokens()) {
+        //BR// This tends to happen when a player loads some save/log file written by the wrong version of a module.
+        //BR// Rather than have a trickle of ABR reports about it, print a more helpful error message.
+        ProblemDialog.showOutdatedModule ("");
```

Comment:
Is there another place this could go, where we catch NoSuchElementException, _outside_ of Decoder? By putting the error handling here, we're tying what is a data processing utility class to the GUI.

Reply:
Hmmm. I feel like SequenceEncoder is implemented & called "a myriad different places"? I might ordinarily suggest putting a try/catch in the various ways of Loading games, but then there's the "load in background" issue. 

Thoughts?

Reply:
Yeah, it's used in... basically every game component there is, which makes dealing with this kind of irritating.

Are there really situations where this exception is recoverable? I've always been under the impression that you're totally screwed when it happens.

Reply:
So yes you're probably basically totally screwed. But we should still display THIS error message instead of the other one. *Possibly* we should exit without allowing the option to continue (keep in mind that the Exception box ALSO lets you try to continue, as inadvisable as that often is). Possibly I should make the message I'm showing even more "no really you're screwed, don't try to keep playing" than I already have.

Reply:
I'm with Joel, I don't think SequecnceEncoder.Decoder is the right place. It has no context to be able to report a meaningful error message.

Since the 2569 error occurs in Decorator.mergeState() which is only called from ChangePiece.executeCommand(), I think it would be better to do the error reporting in ChangePiece. It can actually report the name of the piece that has the problem, and give some guidance as to what should happen. 

When this happens in ChangePiece, we can be pretty sure that the player is trying to run a log file containing ChangePiece Commands generated on an out of date module and can give some advice on this. And display a Stupid User dialog, not the ABR dialog and don't generate an ABR.

If we stick this error handling in SequenceEncoder.Decoder, it could be caused by anything and the error message will have to be so generic as to be useless. The message you are showing for this error only applies to the ChangePiece case, not to any other SE.D problems.

I would vote to put SE.D back to throwing the NoSuchElementException so that we DO get an ABR if this exception happens anywhere else other than ChangePiece.execute. 

You know what we really need is a Version number in each piece's Type that is incremented each time the Type saved in the PieceSlot changes.

Reply:
Okay I *think* I have set this to do what you describe -- please check it out.